test/e2e/support/utils.js

```diff
@@ -475,3 +476,127 @@ export function observeFocusLoss() {
+export function isEmbedding( url ) {
+	return ( request ) => matchURL( 'oembed%2F1.0%2Fproxy' )( request ) && parameterEquals( 'url', url )( request );
+}
+
+/**
+ * Respond to a request with a JSON response.
+ *
+ * @param {string} mockResponse The mock object to wrap in a JSON response.
+ * @return {Promise} Promise that responds to a request with the mock JSON response.
+ */
+export function JSONResponse( mockResponse ) {
+	return async ( request ) => request.respond( getJSONResponse( mockResponse ) );
+}
+
+/**
+ * Creates a function to determine if a request is calling a URL with the substring present.
+ *
+ * @param {string} substring The substring to check for.
+ * @return {function} Function that determines if a request's URL contains substring.
+ */
+export function matchURL( substring ) {
+	return ( request ) => -1 !== request.url().indexOf( substring );
+}
+
+/**
+ * Creates a function to determine if a request has a parameter with a certain value.
+ *
+ * @param {string} parameterName The query parameter to check.
+ * @param {string} value The value to check for.
+ * @return {function} Function that determines if a request's query parameter is the specified value.
+ */
+export function parameterEquals( parameterName, value ) {
+	return ( request ) => {
+		const url = request.url();
+		return value === decodeURIComponent( new RegExp( '.*' + parameterName + '=([^&]+).*' ).exec( url )[ 1 ] );
+	};
+}
+
+/**
+ * Get a JSON response for the passed in object, for use with `request.respond`.
+ *
+ * @param {Object} obj Object to seralise for response.
+ * @return {Object} Response for use with `request.respond`.
+ */
+export function getJSONResponse( obj ) {
+	return {
+		content: 'application/json',
+		body: JSON.stringify( obj ),
+	};
+}
+
+/**
+ * Mocks a request with the supplied mock object, or allows it to run with an optional transform, based on the
+ * deserialised JSON response for the request.
+ *
+ * @param {function} mockCheck function that returns true if the request should be mocked.
+ * @param {Object} mock A mock object to wrap in a JSON response, if the request should be mocked.
+ * @param {function|undefined} responseObjectTransform An optional function that transforms the response's object before the response is used.
+ * @return {Promise} Promise that uses `mockCheck` to see if a request should be mocked with `mock`, and optionally transforms the response with `responseObjectTransform`.
+ */
+export function mockOrTransform( mockCheck, mock, responseObjectTransform = ( obj ) => obj ) {
+	return async ( request ) => {
+		// Because we can't get the responses to requests and modify them on the fly,
+		// we have to make our own request and get the response, then apply the
+		// optional transform to the json encoded object.
+		const response = await fetch(
+			request.url(),
+			{
+				headers: request.headers(),
+				method: request.method(),
+				body: request.postData(),
+			}
+		);
+		const obj = await response.json();
```

Comment:
This variable name is especially confusing because of the inline function using `obj` above; I know they aren't in the same scope or anything but it's a bit confusing to read for me. I had to do a double-take. I think it'd be good to rename all the `obj` with what _kind_ of object you expect them to be if it's at all possible.

Reply:
I've renamed this one `responseObject` because this can be used with more than just embeds, and all we really know for sure it that it's from a response :)